website/blog/2024-01-23-custom-pyfunc/index.md

```diff
@@ -0,0 +1,599 @@
+initialize the model. This can be accomplished with
+[model inference params](https://mlflow.org/docs/latest/models.html#model-inference-params). To use
+inference params, we must pass a valid model signature including `params`. Here’s how to adapt the
+preceding example to use inference params:
+
+```python
+import pandas as pd
+import mlflow
+from mlflow.models import infer_signature
+from mlflow.pyfunc import PythonModel
+
+
+# Custom PythonModel class
+class LinearFunctionInferenceParams(PythonModel):
+    def predict(self, context, model_input, params):
+        """
+        Applies a simple linear transformation
+        to the input data. For example, y = 2x + 3.
+        """
+        slope = params["slope"]
+        # Assuming model_input is a pandas DataFrame with one column
+        return pd.DataFrame(slope * model_input + params["intercept"])
+
+
+# Set default params
+params = {"slope": 2, "intercept": 3}
+
+# Define model signature
+signature = infer_signature(model_input=pd.DataFrame([10, 20, 30]), params=params)
+
+# Save the model with mlflow
+with mlflow.start_run():
+    model_info = mlflow.pyfunc.log_model(
+        artifact_path="model_with_params",
+        python_model=LinearFunctionInferenceParams(),
+        signature=signature,
+    )
+```
+
+After loading the model as before, you can now pass a `params` argument to the `predict` method,
+enabling you to use the same loaded model for different combinations of parameters:
+
+```python
+loaded_model = mlflow.pyfunc.load_model(model_uri=model_info.model_uri)
+
+parameterized_predictions = loaded_model.predict(
+    pd.DataFrame([10, 20, 30]), params={"slope": 2, "intercept": 10}
+)
+print(parameterized_predictions)
+```
+
+```text
+:     0
+: 0  30
+: 1  50
+: 2  70
+```
+
+### Loading external resources with `load_context`
+
+Custom models often require external files such as model weights in order to perform inference.
+These files, or artifacts, must be handled carefully to avoid unnecessarily loading files into
+memory or errors during model serialization. When building custom pyfunc models in MLflow, you can
+use the `load_context` method to handle model artifacts correctly.
+
+The `load_context` method receives a `context` object containing artifacts the model can use during
+inference. You can specify these artifacts using the `artifacts` argument when saving or logging
+models, making them accessible to the `load_context` method via the `context.artifacts` dictionary:
+
+```python
+# Log the model using mlflow with the model file as an artifact
+with mlflow.start_run():
+    model_info = mlflow.pyfunc.log_model(
+        artifact_path="my_custom_model",
+        python_model=<model>,
+        # Specify where artifacts can be found
+        artifacts={
+            "artifact": <artifact_1_path>,
+            "another_artifact": <another_artifact_path>
+        },
+        [...]
```

Comment:
Is this line valid code?

Reply:
Looking at it again, it doesn't do any explanation that the functional code example in the next section shows. I'm going to remove this example, as it's not really useful.